Support to Display Enabled Currencies and relevant icons

thisisayush opened the pull request:
Fixes #130 

![image](https://user-images.githubusercontent.com/22245433/141783064-06ce1d66-20db-49e4-a7bd-89d20b94890e.png)

DarrenWestwood commented on views/templates/front/payment_infos.tpl:
```diff
@@ -0,0 +1,32 @@
+
+<div id="blockonomics-payment-helper"></div>
+<script>
+    let helper = document.getElementById("blockonomics-payment-helper")
```
Adding script to handle the logo height looks like a good solution. We can rename the file to logo_height.tpl
We can also reduce the code as we are targeting a single img
```
var image = document.querySelector("img[src*='blockonomics/views/img']")
image.style.height = "{$blockonomicsLogoHeight}px"
```

thisisayush replied:
Updated. Working fine.

thisisayush commented on blockonomics.php:
```diff
@@ -605,6 +633,27 @@ class="process-icon-refresh"></a>
+
+        $fields_form[2]['form'] = array(
+            'legend' => array(
+                'title' => $this->l('Additional Settings')
```
Updated:

![image](https://user-images.githubusercontent.com/22245433/146316622-7b38b8a6-7858-4532-b0f4-713aa3b22d54.png)

thisisayush commented on blockonomics.php:
```diff
@@ -200,9 +202,26 @@ public function hookPaymentOptions($params)
+        $logoHeight = Configuration::get('BLOCKONOMICS_LOGO_HEIGHT');
+
+        if(!isset($logoHeight)) {
+            $logoHeight = "40";
```
Updated

DarrenWestwood commented on views/templates/front/logo_height.tpl:
```diff
@@ -0,0 +1,23 @@
+
+<script>
+    var image = document.querySelector("img[src*='blockonomics/views/img']")
+    image.style.height = "{$blockonomicsLogoHeight|escape:'htmlall':'UTF-8'}px"
```
We can use {$blockonomicsLogoHeight|intval} here to filter the height.

thisisayush replied:
Done

DarrenWestwood commented on blockonomics.php:
```diff
@@ -200,9 +202,28 @@ public function hookPaymentOptions($params)
+        $active_cryptos = $this->getActiveCurrencies();
+        $logoHeight = Configuration::get('BLOCKONOMICS_LOGO_HEIGHT');
+
+        if (!isset($logoHeight) || !$logoHeight) {
```
This can be removed as using the `|intval` filter in logo_height.tpl will equate BLOCKONOMICS_LOGO_HEIGHT to `0` if this is not set.

thisisayush replied:
Done

DarrenWestwood commented on blockonomics.php:
```diff
@@ -154,6 +154,7 @@ public function installDB()
         Configuration::updateValue('BLOCKONOMICS_TIMEPERIOD', 10);
         Configuration::updateValue('BLOCKONOMICS_BTC', true);
         Configuration::updateValue('BLOCKONOMICS_BCH', false);
+        Configuration::updateValue('BLOCKONOMICS_LOGO_HEIGHT', "40");
```
The default value should be set to 0 here

thisisayush replied:
Done

DarrenWestwood reviewed (commented):
@thisisayush, let me know your thoughts if it is possible to use `setAdditionalInformation` to also update the logo images, as this may allow us to use btc-icon.png and bch-icon.png, without the need for btc-bch-icon.png.
This method will be more scalable in the future, however, might not be required at the moment if it will add too much complexity.

thisisayush commented:
> @thisisayush, let me know your thoughts if it is possible to use `setAdditionalInformation` to also update the logo images, as this may allow us to use btc-icon.png and bch-icon.png, without the need for btc-bch-icon.png. This method will be more scalable in the future, however, might not be required at the moment if it will add too much complexity.

Yes, we can pass the path to image via Smarty Context. Prestashop's Form Helper supports file type as well for configuration. Let me test if we can use that for uploading custom image. Will update you on the same.

thisisayush commented on blockonomics.php:
```diff
@@ -539,6 +560,17 @@ class="process-icon-refresh"></a>
+                    'type'     => 'text',
+                    'label'    => $this->l('Payment Logo Height (px)'),
+                    'desc'     => $this->l(
+                        'Height of logo in Checkout Page in pixels. Default is 0
```
Done

thisisayush commented:
> > @thisisayush, let me know your thoughts if it is possible to use `setAdditionalInformation` to also update the logo images, as this may allow us to use btc-icon.png and bch-icon.png, without the need for btc-bch-icon.png. This method will be more scalable in the future, however, might not be required at the moment if it will add too much complexity.
> 
> Yes, we can pass the path to image via Smarty Context. Prestashop's Form Helper supports file type as well for configuration. Let me test if we can use that for uploading custom image. Will update you on the same.

After researching and trying the basics, yes we can provide the option for the user to upload custom logos, but PS will not manage the file itself, take a look at Banner Module (https://github.com/PrestaShop/ps_banner/blob/dev/ps_banner.php#L127), the uploaded file after being validated needs to be moved, so we can move it to module's directory or someplace else, and we can store that path in the configuration.

DarrenWestwood commented:
> After researching and trying the basics, yes we can provide the option for the user to upload custom logos, but PS will not manage the file itself, take a look at Banner Module (https://github.com/PrestaShop/ps_banner/blob/dev/ps_banner.php#L127), the uploaded file after being validated needs to be moved, so we can move it to module's directory or someplace else, and we can store that path in the configuration.

I do not want the user to upload a logo. I would like to know how much work will be involved in removing btc-bch-icon.png, and instead using btc-icon.png and bch-icon.png?
**Bitcoin only**
```
<img src="/prestashop/modules/blockonomics/views/img/btc-icon.png">
```
**Bitcoin Cash only**
```
<img src="/prestashop/modules/blockonomics/views/img/bch-icon.png">
```
**Both Bitcoin and Bitcoin Cash**
```
<img src="/prestashop/modules/blockonomics/views/img/btc-icon.png"><img src="/prestashop/modules/blockonomics/views/img/bch-icon.png">
```

thisisayush commented:
Not much, we can pass the array of enabled currencies using additional information and use frontend script to add them in place.

DarrenWestwood commented:
> Not much, we can pass the array of enabled currencies using additional information and use frontend script to add them in place.

Okay lets use this method instead, and remove btc-bch-icon.png.

https://github.com/blockonomics/prestashop-plugin/pull/137/files#r770593439 also still needs to be addressed.

thisisayush commented:
> > Not much, we can pass the array of enabled currencies using additional information and use frontend script to add them in place.
> 
> Okay lets use this method instead, and remove btc-bch-icon.png.
> 
> https://github.com/blockonomics/prestashop-plugin/pull/137/files#r770593439 also still needs to be addressed.

To implement this, we'll have to add the helper div back again (https://github.com/blockonomics/prestashop-plugin/pull/137/commits/e66a37d4e283ed6e6adf07c9291ef2b4ca70b30d#diff-3e244bd2586c70ef3e884b2b1a4415d076255a7d46b56bfb5779a611a3743e28R20-R26) to get the target where we want to append the images. I've pushed the new changes with this back to review. It's tested and working. Will apply PS Validator fixes if it looks good.

thisisayush commented on blockonomics.php:
```diff
@@ -689,6 +724,20 @@ public function updateSettings()
             'BLOCKONOMICS_BCH',
             Tools::getValue('BLOCKONOMICS_BCH')
         );
+        Configuration::updateValue(
```
Updated.

DarrenWestwood commented on blockonomics.php:
```diff
@@ -200,9 +202,27 @@ public function hookPaymentOptions($params)
         $offlineOption
-            ->setCallToActionText($this->l('Pay by bitcoin'))
-            ->setLogo(_MODULE_DIR_.'blockonomics/views/img/bitcoin-icon.png')
+            ->setCallToActionText($this->l('Pay by ' . join(' or ', $cryptos)))
```
We can add `->setModuleName('blockonomics')` here, which in turn adds `data-module-name="blockonomics"` to the input checkbox on the checkout. This will then offer a more reliable way to target the correct container. eg
```
let payment_number = document.querySelector("input[data-module-name='blockonomics']").id.split("payment-option-")[1];
let payment_container = document.getElementById("payment-option-" + payment_number + "-container");
{foreach $blockonomicsEnabledLogos as $logo}
    img = document.createElement("img")
    img.src = "{$logo}"
    img.style.height = "{$blockonomicsLogoHeight|intval}px"
    img.style.paddingRight = "5px"
    payment_container.querySelector('label').appendChild(img)
{/foreach}
```

thisisayush replied:
Okay. Let me try.

thisisayush replied:
The query will then become:

```
let container = document.querySelector("input[data-module-name=blockonomics]").parentElement.parentElement.querySelector("label")

{foreach $blockonomicsEnabledLogos as $logo}
    img = document.createElement("img")
    img.src = "{$logo}"
    img.style.height = "{$blockonomicsLogoHeight|intval}px"
    img.style.paddingRight = "5px"
    container.appendChild(img)
{/foreach}
```

DarrenWestwood commented on blockonomics.php:
```diff
@@ -200,9 +202,27 @@ public function hookPaymentOptions($params)
-            ->setLogo(_MODULE_DIR_.'blockonomics/views/img/bitcoin-icon.png')
+            ->setCallToActionText($this->l('Pay by ' . join(' or ', $cryptos)))
+            ->setAdditionalInformation(
+                $this->context->smarty->fetch('module:blockonomics/views/templates/front/logo_height.tpl')
```
The file location can be moved to the more appropriate module:blockonomics/views/templates/**hook**/logo_height.tpl, as this is used in hookPaymentOptions.

thisisayush replied:
Okay

DarrenWestwood commented:
views/img/btc-bch-icon.png should be removed, and btc-icon.png/bch-icon.png updated to use .svg images.

DarrenWestwood commented:
`.querySelector("label")` is not working correctly with some themes. We may be better off using setLogo, to set the default to btc-icon.png and then replacing this image with the active images
```
var image = document.querySelector("img[src*='blockonomics/views/img']");
var image_container = document.createElement("span")
{foreach $blockonomicsEnabledLogos as $logo}
    img = document.createElement("img")
    img.src = "{$logo|escape:'htmlall':'UTF-8'}"
    img.style.height = "{$blockonomicsLogoHeight|intval}px"
    img.style.paddingRight = "5px"
    image_container.append(img);
{/foreach}
image.parentNode.replaceChild(image_container, image);
```

Using `setAdditionalInformation` is also creating an empty information area when the checkbox is selected. Is there any way to prevent this?
![Prestashop(1)](https://user-images.githubusercontent.com/39666372/147127511-21e3b1ec-f3a0-4df0-8a9e-e93aaf33147a.png)

thisisayush commented:
> `.querySelector("label")` is not working correctly with some themes. We may be better off using setLogo, to set the default to btc-icon.png and then replacing this image with the active images
> 
> ```
> var image = document.querySelector("img[src*='blockonomics/views/img']");
> var image_container = document.createElement("span")
> {foreach $blockonomicsEnabledLogos as $logo}
>     img = document.createElement("img")
>     img.src = "{$logo|escape:'htmlall':'UTF-8'}"
>     img.style.height = "{$blockonomicsLogoHeight|intval}px"
>     img.style.paddingRight = "5px"
>     image_container.append(img);
> {/foreach}
> image.parentNode.replaceChild(image_container, image);
> ```
> 
> Using `setAdditionalInformation` is also creating an empty information area when the checkbox is selected. Is there any way to prevent this? ![Prestashop(1)](https://user-images.githubusercontent.com/39666372/147127511-21e3b1ec-f3a0-4df0-8a9e-e93aaf33147a.png)

We can hide it using CSS. Let me try something.

thisisayush commented:
Updated the commit. Here's how it looks in Dev:

When Payment Mode is selected as Blokconomics: 
![image](https://user-images.githubusercontent.com/22245433/147215937-6d14966c-8b02-4d18-8536-efa459f159d9.png)

![image](https://user-images.githubusercontent.com/22245433/147216035-041b04de-270f-460b-8ecf-ed00844ff329.png)


Also added some properties to the logos parent container to align the logos with some themes.

blockonomics commented:
Tested looking good. Can you do the following changes that will update the text
```

diff --git a/blockonomics.php b/blockonomics.php
index c6a91e6..7830d5b 100755
--- a/blockonomics.php
+++ b/blockonomics.php
@@ -565,10 +565,10 @@ class Blockonomics extends PaymentModule
                 ),
                 array(
                     'type'     => 'text',
-                    'label'    => $this->l('Payment Logo Height (px)'),
+                    'label'    => $this->l('Pay by bitcoin icon size'),
                     'desc'     => $this->l(
-                        'Height of logo in Checkout Page in pixels.
-                         Set to 0 to disable logo.'
+                        'Size in pixels.
+                         Set 0 to disable icon'
                     ),
                     'name'     => 'BLOCKONOMICS_LOGO_HEIGHT',
                     'required' => false,

```

thisisayush commented:
> Tested looking good. Can you do the following changes that will update the text
> 
> ```
> 
> diff --git a/blockonomics.php b/blockonomics.php
> index c6a91e6..7830d5b 100755
> --- a/blockonomics.php
> +++ b/blockonomics.php
> @@ -565,10 +565,10 @@ class Blockonomics extends PaymentModule
>                  ),
>                  array(
>                      'type'     => 'text',
> -                    'label'    => $this->l('Payment Logo Height (px)'),
> +                    'label'    => $this->l('Pay by bitcoin icon size'),
>                      'desc'     => $this->l(
> -                        'Height of logo in Checkout Page in pixels.
> -                         Set to 0 to disable logo.'
> +                        'Size in pixels.
> +                         Set 0 to disable icon'
>                      ),
>                      'name'     => 'BLOCKONOMICS_LOGO_HEIGHT',
>                      'required' => false,
> ```

Done

blockonomics commented:
@DarrenWestwood  if you approve this, we can merge and do a release

DarrenWestwood reviewed (approved):
The changes are neat and working well.
